### PR TITLE
sql/parser: normalize after type checking

### LINKED ISSUE #2638: sql: NormalizeExpr and TypeCheckExpr are tangled
It's not clear which function should be run first, and the application of these functions does not commute. For instance, NormalizeExpr does this:
```sql
before: CASE 1 WHEN 1 THEN 'one' WHEN 2 THEN 2 END
after: 'one'
```
Note that this expression does not type-check.

On the other hand, TypeCheckExpr contains the following snippet:
```go
	case Row:
		// NormalizeExpr transforms this into Tuple.

	case *ParenExpr:
		// NormalizeExpr unwraps this.

	case *ComparisonExpr:
		return typeCheckComparisonExpr(t)

	case *RangeCond:
		// NormalizeExpr transforms this into an AndExpr.
```

### PR DESCRIPTION
**Pull request by tamird**
Fixes #2638.

Pretty much a freebie after the work of moving to methods.

### CODE REVIEW

File: sql/parser/normalize.go [hunk 1]
```diff
@@ -97,8 +97,11 @@ func (expr *ComparisonExpr) normalize(v *normalizeVisitor) Expr {
 				// The left side is const and the right side is a binary expression or a
 				// variable. Flip the comparison op so that the right side is const and
 				// the left side is a binary expression or variable.
-				expr.Operator = invertComparisonOp(expr.Operator)
-				expr.Left, expr.Right = expr.Right, expr.Left
+				*expr = ComparisonExpr{
+					Operator: invertComparisonOp(expr.Operator),
+					Left:     expr.Right,
+					Right:    expr.Left,
+				}
```
Comment (petermattis) [inline sql/parser/normalize.go]: This is equivalent to the previous code. Any reason for the change?

File: sql/parser/normalize.go [hunk 2]
```diff
@@ -119,19 +122,25 @@ func (expr *ComparisonExpr) normalize(v *normalizeVisitor) Expr {
...
+					left.Operator = Mult
+				default:
+					rotating = false
+				}
+
+				if rotating {
+					// Clear the function caches since we're rotating.
+					expr.fn.fn = nil
+					left.fn.fn = nil
+
```
Comment (petermattis) [inline sql/parser/normalize.go]: Ditto about equivalence. I don't find this any more (or less) readable. Wondering why bother.

**Comment by petermattis** [comment]
Was there a reason for performing normalization before type checking before? Was it just to not have to worry about the difference between `Row` and `Tuple`?

File: sql/parser/type_check.go [hunk 3]
```diff
@@ -30,12 +30,12 @@ import (
 // NormalizeExpr or TypeCheck() return one, and otherwise returns
 // the Expr returned by NormalizeExpr.
 func (ctx EvalContext) NormalizeAndTypeCheckExpr(expr Expr) (Expr, error) {
```
Comment (petermattis) [inline sql/parser/type_check.go]: Rename to `TypeCheckAndNormalizeExpr`?

File: sql/parser/normalize.go (hunk 1 above)
Comment (tamird) [inline sql/parser/normalize.go]: it's not equivalent - the function cache is cleared here, but was not previously

**Comment by tamird** [comment]
> Was there a reason for performing normalization before type checking before? Was it just to not have to worry about the difference between `Row` and `Tuple`?

Yep.

File: sql/parser/normalize.go (hunk 1 above)
Comment (petermattis) [inline sql/parser/normalize.go]: Oh, that's subtle. Deserves a comment as the other clearing of the function caches have.

File: sql/parser/normalize.go (hunk 2 above)
Comment (tamird) [inline sql/parser/normalize.go]: Yeah, this is the second commit, and it is equivalent. I find this easier to read (and I wanted to similarly refactor the other case below, but couldn't figure it out.

I'll leave it out for now.

**Comment by petermattis** [comment]
LGTM

File: sql/parser/normalize.go (hunk 1 above)
Comment (tamird) [inline sql/parser/normalize.go]: done

File: sql/parser/type_check.go (hunk 3 above)
Comment (tamird) [inline sql/parser/type_check.go]: done

**Pull request merged**